src/Tizen.System.Session/Session/SessionStructs.cs

```diff
@@ -0,0 +1,234 @@
+        public byte userName_15;
+        public byte userName_16;
+        public byte userName_17;
+        public byte userName_18;
+        public byte userName_19;
+
+        public string userName
+        {
+            get
+            {
+                return UnrolledBytesToStringConverter.Convert20(
+                    userName_00, userName_01, userName_02, userName_03,
+                    userName_04, userName_05, userName_06, userName_07,
+                    userName_08, userName_09, userName_10, userName_11,
+                    userName_12, userName_13, userName_14, userName_15,
+                    userName_16, userName_17, userName_18, userName_19
+                );
+            }
+        }
+    }
+
+    [StructLayout(LayoutKind.Sequential, CharSet = CharSet.Ansi)]
+    internal struct RemoveUser
+    {
+        public byte userName_00;
+        public byte userName_01;
+        public byte userName_02;
+        public byte userName_03;
+        public byte userName_04;
+        public byte userName_05;
+        public byte userName_06;
+        public byte userName_07;
+        public byte userName_08;
+        public byte userName_09;
+        public byte userName_10;
+        public byte userName_11;
+        public byte userName_12;
+        public byte userName_13;
+        public byte userName_14;
+        public byte userName_15;
+        public byte userName_16;
+        public byte userName_17;
+        public byte userName_18;
+        public byte userName_19;
+        public string userName
+        {
+            get
+            {
+                return UnrolledBytesToStringConverter.Convert20(
+                    userName_00, userName_01, userName_02, userName_03,
+                    userName_04, userName_05, userName_06, userName_07,
+                    userName_08, userName_09, userName_10, userName_11,
+                    userName_12, userName_13, userName_14, userName_15,
+                    userName_16, userName_17, userName_18, userName_19
+                );
+            }
+        }
+    }
+    [StructLayout(LayoutKind.Sequential, CharSet = CharSet.Ansi, Pack = 4)]
+    internal struct SwitchUser
+    {
+        public Int64 switchID;
+        public byte userNamePrev_00;
+        public byte userNamePrev_01;
+        public byte userNamePrev_02;
+        public byte userNamePrev_03;
+        public byte userNamePrev_04;
+        public byte userNamePrev_05;
+        public byte userNamePrev_06;
+        public byte userNamePrev_07;
+        public byte userNamePrev_08;
+        public byte userNamePrev_09;
+        public byte userNamePrev_10;
+        public byte userNamePrev_11;
+        public byte userNamePrev_12;
+        public byte userNamePrev_13;
+        public byte userNamePrev_14;
+        public byte userNamePrev_15;
+        public byte userNamePrev_16;
+        public byte userNamePrev_17;
+        public byte userNamePrev_18;
+        public byte userNamePrev_19;
+
```

Comment:
You don't need it
Please check it
https://learn.microsoft.com/en-us/dotnet/standard/native-interop/customize-struct-marshalling#customizing-string-field-marshalling

https://github.com/Samsung/TizenFX/blob/d586fc850c80f9db2004caef3bedc5b939bfed85/src/Tizen.Multimedia.Camera/Interop/Interop.Camera.cs#L361-L362

Reply:
Thanks for hints. Unfortunately this case is more complicated, as the structure actually contains unions, which in turn constist of another structures:

    typedef struct subsession_event_info {
    
        subsession_event_type_e event;
        int session_uid;
    
        union {
            struct {
                subsession_user_t user;
            } add_user;
    
            struct {
                subsession_user_t user;
            } remove_user;
        
            struct {
                uint64_t switch_id;
                subsession_user_t prev_user;
                subsession_user_t next_user;
            } switch_user;
        };
    } subsession_event_info;

We've tried to use LayoutKind Explicit for unions and LayoutKind.Sequential for structs, but this unfortunately didn't work properly, event with a proper explicitly stated alignments.

Reply:
here union sample code
https://learn.microsoft.com/en-us/dotnet/framework/interop/marshalling-classes-structures-and-unions#unions-sample

Reply:
A union/several unions are not a problem if they occur separately. The problem is when they are members of a struct. It's the situation that we have with `libsessiond` (please see the attached definition of the C struct in the above comment). It leads to a runtime error: "object-field at offset 4 is incorrectly aligned or overlapped by a non-object field" when marshalled with StructLayout(LayoutKind.Explicit) and FieldOffset. A very similar situation was described here:

https://stackoverflow.com/questions/71426822/marshaling-complex-c-structs-with-c-unions-for-c-sharp

Reply:
Maybe you can marshal with it
```
    internal struct SubsessionEventInfoNative
    {
        [FieldOffset(0)]
        public SessionEventType eventType;
        [FieldOffset(4)]
        public int sessionUID;
        [FieldOffset(8)]
        [MarshalAs(UnmanagedType.LPArray, SizeConst = 20)]
        public string AddUser;
        [FieldOffset(8)]
        [MarshalAs(UnmanagedType.LPArray, SizeConst = 20)]
        public string RemoveUser;
        [FieldOffset(8)]
        public Int64 switchID;
        [MarshalAs(UnmanagedType.LPArray, SizeConst = 20)]
        public string PrevUser;
        [MarshalAs(UnmanagedType.LPArray, SizeConst = 20)]
        public string NextUser;
    }

```

Reply:
Anyway, .net marshal api can convert any struct/union on unmanaged, I hope to you check more marshal API

Reply:
``` c#
    internal struct SubsessionEventInfoNative
    {
        [FieldOffset(0)]
        public SessionEventType eventType;
        [FieldOffset(4)]
        public int sessionUID;
        [FieldOffset(8)]
        public IntPtr AddUserPtr;
        [FieldOffset(8)]
        public IntPtr RemoveUserPtr;
        [FieldOffset(8)]
        public Int64 switchID;
        [FieldOffset(16)]
        public IntPtr PrevUserPtr;
        [FieldOffset(36)]
        public IntPtr NextUserPtr;
    }
```
And It also works like it
``` c#
  var eventInfo = (SubsessionEventInfoNative)Marhsal.PtrToStructure( ...)
  var adduser = Marhsal.PtrToStringAnsi(eventInfo.AddUserPtr, 20);
...
  var prevUser = Marhsal.PtrToStringAnsi(eventInfo.PrevUserPtr, 20);

```

Reply:
@a-michalski2 Please take a look if you can access CODE.
https://github.sec.samsung.net/dotnet/home/issues/421#issuecomment-1686515
Camera API marshals way more complicated struct than this one.

Reply:
Thanks for suggestions. Unfortunately the following doesn't work (PrevUserPtr always contains null). 

``` C#
    internal struct SubsessionEventInfoNative
    {
        [FieldOffset(0)]
        public SessionEventType eventType;
        [FieldOffset(4)]
        public int sessionUID;

        [FieldOffset(8)]
        public IntPtr AddUserPtr;

        [FieldOffset(8)]
        public IntPtr RemoveUserPtr;

        [FieldOffset(8)]
        public Int64 switchID;
        [FieldOffset(16)]
        public IntPtr PrevUserPtr;
        [FieldOffset(36)]
        public IntPtr NextUserPtr;
    }
```

We've come up with something though:

``` C#
    [StructLayout(LayoutKind.Sequential, CharSet = CharSet.Ansi)]
    internal struct AddUserStruct
    {
        [MarshalAs(UnmanagedType.LPArray, SizeConst = 20)]
        internal string userName;
    }
    [StructLayout(LayoutKind.Sequential, CharSet = CharSet.Ansi)]
    internal struct RemoveUserStruct
    {
        [MarshalAs(UnmanagedType.LPArray, SizeConst = 20)]
        internal userName;
    }
    [StructLayout(LayoutKind.Sequential, CharSet = CharSet.Ansi)]
    internal struct SwitchUserStruct
    {
        internal Int64 switchID;
        [MarshalAs(UnmanagedType.LPArray, SizeConst = 20)]
        internal string prevUser;
        [MarshalAs(UnmanagedType.LPArray, SizeConst = 20)]
        internal string nextUser;
    }

    [StructLayout(LayoutKind.Explicit)]
    internal struct UserDataStruct
    {
        [FieldOffsetAttribute(0)]
        internal AddUserStruct AddUserData;
        [FieldOffsetAttribute(0)]
        internal RemoveUserStruct RemoveUserData;
        [FieldOffsetAttribute(0)]
        internal SwitchUserStruct SwitchUserData;
    }

    [StructLayout(LayoutKind.Sequential)]
    internal struct SessionEventInfo
    {
        internal SessionEventType eventType;
        internal int SessionUID;
        internal UserDataStruct userData;
    }
```

We'll be testing this today to see if it works.

Reply:
Unfortunately, this solution doesn't work either.

It truncates first 8 bytes of the `AddUserStruct.userName` and `RemoveUserStruct.userName` (`SwitchUserStruct.prevUser` and `SwitchUserStruct.nextUser` seem to be properly marshalled):
```
    Adding test users...
    Added - �
    Added - e
    Added - rf
    Added - murf
    Added - murf
    Added - murf
    Added - smurf
    Added - urf
    Added - urf
    Added - rf
    Users added
    Curr 
    Curr baby_smurf
    Curr ㅋㅌㅍ
    Curr smurfette
    Curr brainy_smurf
    Curr clumsy_smurf
    Curr greedy_smurf
    Curr grouchy_smurf
    Curr hefty_smurf
    Curr jokey_smurf
    Curr lazy_smurf
    Switching to each user...
    Switched from 
    Switched from ㅋㅌㅍ
    Switched from smurfette
    Switched from baby_smurf
    Switched from brainy_smurf
    Switched from clumsy_smurf
    Switched from greedy_smurf
    Switched from grouchy_smurf
    Switched from hefty_smurf
    Switched from jokey_smurf
    Switching finished.
```
Another attempt:
``` C#
 [NativeStruct("subsession_event_info", Include="sessiond.h", PkgConfig="libsessiond")]
    [StructLayout(LayoutKind.Explicit)]
    internal struct SubsessionEventInfoNative
    {
        [FieldOffset(0)]
        public SessionEventType eventType;
        [FieldOffset(4)]
        public int sessionUID;

        [FieldOffset(8)]
        [MarshalAs(UnmanagedType.ByValArray, SizeConst = 20)]
        public char[] AddUser;

        [FieldOffset(8)]
        [MarshalAs(UnmanagedType.ByValArray, SizeConst = 20)]
        public char[] RemoveUser;

        [FieldOffset(8)]
        public Int64 SwitchID;

        [FieldOffset(16)]
        [MarshalAs(UnmanagedType.ByValArray, SizeConst = 20)]
        public char[] SwitchUserPrev;
        [FieldOffset(36)]
        [MarshalAs(UnmanagedType.ByValArray, SizeConst = 20)]
        public char[] SwitchUserNext;
    }
```
gives the following runtime error:
```
Unhandled exception. System.TypeLoadException: Could not load type 'Tizen.System.SubsessionEventInfoNative' from assembly 'Tizen.System.Session, Version=4.0.0.0, Culture=neutral, PublicKeyToken=eba370b203a2e242' because it contains an object field at offset 8 that is incorrectly aligned or overlapped by a non-object field.
```

We've tried several other possible solutions but we end up with either the runtime error or a truncation of the first 4 or 8 characters of the `userName` for `AddUser` and `RemoveUser` union.

Reply:
Try it also
``` c#
IntPtr nativePtr = ...
string addUser = Marshal.PtrToStringAnsi(IntPtr.Add(nativePtr, 8), 20)
...
string removeUser = Marshal.PtrToStringAnsi(IntPtr.Add(nativePtr, 8), 20)
...
string prevUser = Marshal.PtrToStringAnsi(IntPtr.Add(nativePtr, 16), 20)
string nextUser = Marshal.PtrToStringAnsi(IntPtr.Add(nativePtr, 36), 20)
```

Reply:
Thanks for further suggestions. We'll be trying to test more possible solutions. We're now moving the code to produce a MRE consisting only of the `.c` file with structures and a mocked function filling it and a simple C# project with a single `.cs` file to be able to build and check it quickly on Ubuntu without having to build the whole `TizenFX` which takes a lot of time.

Reply:
We finally managed to get this struct working without using unrolled arrays. Thanks for your help!